In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
data = pd.read_csv('drive/My Drive/data.csv')
macro = pd.read_csv('drive/My Drive/data_macro.csv')

### <center> Задание:

1. Данные представлены двумя файлами, для начала их нужно объединить в один датасет 
2. Необходимо провести анализ и очистку "грязных данных", найти аномалии и выбросы, подумать, как их лучше обработать (скорректировать значения, удалить, оставить и обработать отдельно). Для анализа данных и поиска аномалий рекомендуется использовать в том числе и визуализацию
3. В данных есть пропуски, подумать, как их лучше обработать. Если заполнять, то чем именно лучше заполнить в каждом отдельном случае
4. В конце ноутбука заполнить ячейку **"Отчет по анализу"**, в которой по каждому столбцу описать, ***какие именно аномалии были в нем найдены*** (например, выбросы, пропуски, логичесткие противоречия и т.д.) и как ***обработаны*** (удалены, заменены другим значением и т.д.).  
    
**Важно!**
Нужно постараться **найти и исправить** максимальное количество "косяков" в данных, чтобы на выходе получить чистый набор данных, с которым можно дальше работать. Важно не просто найти, но и обработать "косяк", если он не обработан, то зачтен не будет.

**Максимальное количество баллов за работу - 26**

**Описание данных:**

***data.csv*** 
- **id:** id транзакции
- **timestamp:** дата продажи (транзакции)
- **full_sq:** общая площадь 
- **life_sq:** жилая площадь 
- **floor:** этаж 
- **max_floor:** количество этажей в здании
- **material:** материал, из которого изговолены стены 
- **build_year:** год строительства
- **num_room:** количество жилых комнат
- **kitch_sq:** площадь кухни
- **full_all:** количество населения в регионе
- **state:** жилищные условия
- **sub_area:** название территории
- **price_doc:** цена квартиры (целевая переменная)

***macro.csv***
- **timestamp** : дата, на которую актуальны макроэкономические показатели
- **salary** : средняя зарплата в регионе
- **fixed_basket**: стоимость потребительской корзины
- **rent_price_3room_eco**: стоимость аренды 3-х комнатного жилья эконом-класса
- **rent_price_2room_eco**: стоимость аренды 2-х комнатного жилья эконом-класса
- **rent_price_1room_eco**: стоимость аренды 1-но комнатного жилья эконом-класса
- **average_life_exp**: средняя продолжительность жизни в регионе

In [ ]:
df = data.merge(macro, how="left")
df

id   timestamp  ...  rent_price_1room_eco  average_life_exp
0       8059  2013-05-21  ...                  2.31             76.37
1       8138  2013-05-25  ...                  2.31             76.37
2       8156  2013-05-27  ...                  2.31             76.37
3       8157  2013-05-27  ...                  2.31             76.37
4       8178  2013-05-28  ...                  2.31             76.37
...      ...         ...  ...                   ...               ...
18861  21562  2014-08-15  ...                 34.83             76.70
18862   4796  2012-11-19  ...                 32.93             75.74
18863    466  2011-11-25  ...                 31.53             75.79
18864  27416  2015-01-21  ...                 33.88             76.77
18865   1723  2012-03-29  ...                 31.64             75.74

[18866 rows x 20 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18866 entries, 0 to 18865
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    18866 non-null  int64  
 1   timestamp             18866 non-null  object 
 2   full_sq               18866 non-null  int64  
 3   life_sq               17291 non-null  float64
 4   floor                 18838 non-null  float64
 5   state                 15505 non-null  float64
 6   max_floor             17445 non-null  float64
 7   material              17445 non-null  float64
 8   build_year            16866 non-null  object 
 9   num_room              17445 non-null  float64
 10  kitch_sq              17445 non-null  float64
 11  full_all              18866 non-null  int64  
 12  sub_area              18866 non-null  object 
 13  price_doc             18866 non-null  int64  
 14  salary                18866 non-null  float64
 15  fixed_basket       

In [ ]:
df.describe()

id       full_sq  ...  rent_price_1room_eco  average_life_exp
count  18866.000000  18866.000000  ...          18866.000000      18866.000000
mean   18799.788614     53.940899  ...             33.376801         76.599302
std     7082.428695     44.116185  ...              2.346476          0.229903
min        9.000000      0.000000  ...              2.310000         75.740000
25%    13513.250000     38.000000  ...             32.960000         76.370000
50%    19043.500000     49.000000  ...             33.650000         76.700000
75%    24748.750000     63.000000  ...             34.350000         76.700000
max    30473.000000   5326.000000  ...             35.370000         76.770000

[8 rows x 17 columns]

# full_sq

In [ ]:
df['full_sq'].unique()

array([ 30,  53,  77,  45,  38,  63,  74,  93,  51,  43,  54,  44,  32,
        52,  56,  31,  46,  75,  41,  59,  37,  40,  33,  39,  60,  71,
       114,  58,  66,  42,  36,  78,  61,  35,  57,  65,  64,  48,  50,
        49,  79, 142,  55,  34,  83,  81, 106, 123, 100,  90,  95,  85,
        47,  87,  67,  73,  92,  84,  99,  62,  76,  68,  80,  82, 119,
        97,  86,  69, 125, 101, 138, 118, 121,  72, 139, 133,  89,  94,
        96, 104, 132, 126, 155, 116, 102, 143, 117,  91, 112,  70, 113,
       165, 107, 122, 144, 202, 134,  98, 120, 110, 167, 124, 115, 137,
       108, 109,  88, 160, 105, 111, 131, 103, 183, 215, 174, 146, 127,
       135, 153, 189, 172, 186, 220, 181, 141, 161, 219, 148, 154, 136,
       129, 275, 209, 177, 145, 150, 184, 206, 128, 182, 159, 169, 176,
       151, 637, 218, 216, 166, 207, 157, 179, 199, 407, 187, 170, 149,
       195, 197, 130, 211, 158, 185, 156, 326, 303, 412, 193])

In [ ]:
df[df['full_sq'] == df['full_sq'].max()]

id   timestamp  ...  rent_price_1room_eco  average_life_exp
18428  3530  2012-09-07  ...                  32.9             75.74

[1 rows x 20 columns]

In [ ]:
df['full_sq'].loc[18428] = 326

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [ ]:
df.describe()

id       full_sq  ...  rent_price_1room_eco  average_life_exp
count  18866.000000  18866.000000  ...          18866.000000      18866.000000
mean   18799.788614     53.675872  ...             33.376801         76.599302
std     7082.428695     21.834569  ...              2.346476          0.229903
min        9.000000      0.000000  ...              2.310000         75.740000
25%    13513.250000     38.000000  ...             32.960000         76.370000
50%    19043.500000     49.000000  ...             33.650000         76.700000
75%    24748.750000     63.000000  ...             34.350000         76.700000
max    30473.000000    637.000000  ...             35.370000         76.770000

[8 rows x 17 columns]

In [ ]:
for i in range(0, 18866, 1):
  if df['full_sq'].loc[i] < 30:
    df['full_sq'].loc[i] = 30
df['full_sq'].unique()

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


array([ 30,  53,  77,  45,  38,  63,  74,  93,  51,  43,  54,  44,  32,
        52,  56,  31,  46,  75,  41,  59,  37,  40,  33,  39,  60,  71,
       114,  58,  66,  42,  36,  78,  61,  35,  57,  65,  64,  48,  50,
        49,  79, 142,  55,  34,  83,  81, 106, 123, 100,  90,  95,  85,
        47,  87,  67,  73,  92,  84,  99,  62,  76,  68,  80,  82, 119,
        97,  86,  69, 125, 101, 138, 118, 121,  72, 139, 133,  89,  94,
        96, 104, 132, 126, 155, 116, 102, 143, 117,  91, 112,  70, 113,
       165, 107, 122, 144, 202, 134,  98, 120, 110, 167, 124, 115, 137,
       108, 109,  88, 160, 105, 111, 131, 103, 183, 215, 174, 146, 127,
       135, 153, 189, 172, 186, 220, 181, 141, 161, 219, 148, 154, 136,
       129, 275, 209, 177, 145, 150, 184, 206, 128, 182, 159, 169, 176,
       151, 637, 218, 216, 166, 207, 157, 179, 199, 407, 187, 170, 149,
       195, 197, 130, 211, 158, 185, 156, 326, 303, 412, 193])

# life_sq


In [ ]:
df['life_sq'].unique()

array([ 15.,  30.,  41.,  27.,  20.,  46.,  92.,  50.,  19.,  29.,  31.,
        28.,  18.,  33.,  21.,  45.,  24.,  32.,  38.,  42.,  23.,  69.,
        25.,  43.,  35.,  17.,  49., 132.,  40.,  44.,  70.,  54.,  72.,
        37.,  48.,  39.,  52.,  16.,  34.,  64.,  67.,  56.,  36.,  63.,
        22.,  53.,  55.,  58.,  26.,  47.,  51.,  60.,  79.,  77.,  75.,
        66.,  84.,  76., 124.,  57.,  81.,  59.,  61.,  78.,  62.,  88.,
       112.,  68., 114.,  65., 110.,  73., 125., 105., 150., 121.,  93.,
        94., 115., 119.,  86.,  74.,  85.,  97., 123.,  82.,  71.,  80.,
        87.,  89.,  91., 122.,  98., 102.,  99., 166., 103., 134.,  90.,
       118.,  96., 116., 104.,  83., 157., 176., 145., 147., 232., 100.,
       194., 117., 139.,  95., 109., 146., 101., 106., 108., 627., 111.,
       174., 113., 120., 131., 107., 152., 155., 133., 164., 397., 144.,
       172., 162., 178., 159.])

In [ ]:
df['life_sq'] = df['life_sq'].fillna(0)
df[df['life_sq']==0]

Empty DataFrame
Columns: [id, timestamp, full_sq, life_sq, floor, state, max_floor, material, build_year, num_room, kitch_sq, full_all, sub_area, price_doc, salary, fixed_basket, rent_price_3room_eco, rent_price_2room_eco, rent_price_1room_eco, average_life_exp]
Index: []

In [ ]:
for i in range(0, 18866, 1):
  if df['life_sq'].loc[i] < 15:
    df['life_sq'].loc[i] = 15
  if df['life_sq'].loc[i] >= df['full_sq'].loc[i]:
    df['life_sq'].loc[i] = df['full_sq'].loc[i] - df['kitch_sq'].loc[i]

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [ ]:
vals = [i for i in df['life_sq'].unique() if isinstance(i, float)]
vals

# kitch_sq

In [ ]:
df['kitch_sq'].unique()

array([15., 16., 25., 19., 17., 21., 20., 18., 22., 24., 23.])

In [ ]:
df[df['kitch_sq']>=1000]

id   timestamp  ...  rent_price_1room_eco  average_life_exp
663    10371  2013-09-10  ...                 34.53             76.37
9711   21418  2014-08-11  ...                 34.83             76.70
15530  28737  2015-03-31  ...                 31.35             76.77

[3 rows x 20 columns]

In [ ]:
df['kitch_sq'] = df['kitch_sq'].fillna(15)

In [ ]:
for i in range (0, 18866, 1):
  if df['kitch_sq'].loc[i] < 15:
    df['kitch_sq'].loc[i] = 15
  elif df['kitch_sq'].loc[i] > 25:
    df['kitch_sq'].loc[i] = 25

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


# max_floor

In [ ]:
df['max_floor'].unique()

array([ 5., 16., 17.,  9., 24.,  1., 10., 25., 14., 12.,  8., 18., 23.,
       20.,  6., 22.,  7.,  4., 30., 11.,  3., 40., 32., 15., 27.,  2.,
       19., 28., 29., 39., 38., 13., 31., 21., 26., 37., 33., 47., 36.,
       34., 48., 42., 57., 44., 45., 41.])

In [ ]:
df[df['max_floor'].isna()==True]

id   timestamp  ...  rent_price_1room_eco  average_life_exp
16866  2971  2012-07-24  ...                 31.10             75.74
16867  3217  2012-08-15  ...                 32.11             75.74
16868   386  2011-11-17  ...                 31.53             75.79
16871  5289  2012-12-15  ...                 33.10             75.74
16874  8136  2013-05-25  ...                  2.31             76.37
...     ...         ...  ...                   ...               ...
18857  9074  2013-07-09  ...                 32.55             76.37
18858   554  2011-12-06  ...                 30.74             75.79
18862  4796  2012-11-19  ...                 32.93             75.74
18863   466  2011-11-25  ...                 31.53             75.79
18865  1723  2012-03-29  ...                 31.64             75.74

[1421 rows x 20 columns]

In [ ]:
df['max_floor'] = df['max_floor'].fillna(0)

In [ ]:
for i in range(0, 18866, 1):
  if (df['max_floor'].loc[i] == 0) or (pd.isna(df['floor'].loc[i])):
    df['max_floor'].loc[i] = df['max_floor'].median()

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


# floor

In [ ]:
df['floor'].unique()

array([ 2., 10.,  6., 15., 17., 12.,  3.,  7.,  1., 13.,  5.,  9., 11.,
       16.,  4.,  8., 14., 20., 18., 24., 19., 25., 22., 21., 23., 37.,
       30., 32., 28., 26., 31., 36., 34., 41., 77., 27., 29., 44., 33.])

In [ ]:
df[df['floor'].isna()==True]

Empty DataFrame
Columns: [id, timestamp, full_sq, life_sq, floor, state, max_floor, material, build_year, num_room, kitch_sq, full_all, sub_area, price_doc, salary, fixed_basket, rent_price_3room_eco, rent_price_2room_eco, rent_price_1room_eco, average_life_exp]
Index: []

In [ ]:
df['floor'] = df['floor'].fillna(0)

In [ ]:
for i in range(0, 18866, 1):
  if (df['floor'].loc[i] == 0) and (df['max_floor'].loc[i]>df['floor'].median()):
    df['floor'].loc[i] = df['floor'].median()
  elif (df['floor'].loc[i] == 0 and df['max_floor'].loc[i]<=df['floor'].median()) or df['floor'].loc[i]>df['max_floor'].loc[i]:
    df['floor'].loc[i] = df['max_floor'].loc[i]

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [ ]:
df[df['floor']>df['max_floor']]

Empty DataFrame
Columns: [id, timestamp, full_sq, life_sq, floor, state, max_floor, material, build_year, num_room, kitch_sq, full_all, sub_area, price_doc, salary, fixed_basket, rent_price_3room_eco, rent_price_2room_eco, rent_price_1room_eco, average_life_exp]
Index: []

# state

In [ ]:
df['state'].unique()

array([3., 1., 2., 4.])

In [ ]:
df['state'] = df['state'].fillna(0)

In [ ]:
df[df['state']==33]

id   timestamp  ...  rent_price_1room_eco  average_life_exp
482  10092  2013-08-27  ...                 33.15             76.37

[1 rows x 20 columns]

In [ ]:
df['state'].loc[482] = 3

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [ ]:
for i in range(0, 18866, 1):
  if (df['state'].loc[i] == 0):
    df['state'].loc[i] = df['state'].median()

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


# material

In [ ]:
df['material'].unique()

array([2., 1., 6., 4., 5., 3.])

In [ ]:
df['material'] = df['material'].fillna(df['material'].median())

# build_year

In [ ]:
# находим строковые значения в датафрейме
str_values = [i for i in df['build_year'].unique() if isinstance(i, str)]
str_values 
# ind_to_inplace = df[df['MasVnrArea'].isin(str_values)].index
# df.loc[ind_to_inplace,'MasVnrArea'] = 0
# # также заполним сразу единственный пропуск
# df['MasVnrArea'] = df['MasVnrArea'].fillna(0)


['1907.0',
 '1980.0',
 '2014.0',
 '1970.0',
 '1982.0',
 '2013.0',
 '2004.0',
 '2003.0',
 '1957.0',
 '1986.0',
 '1960.0',
 '1995.0',
 '1979.0',
 '1975.0',
 '1987.0',
 '1962.0',
 '1969.0',
 '1993.0',
 '1996.0',
 '1972.0',
 '2011.0',
 '1965.0',
 '2010.0',
 '1985.0',
 '2006.0',
 '1961.0',
 '1971.0',
 '1978.0',
 '1966.0',
 '1967.0',
 '2000.0',
 '1964.0',
 '1977.0',
 '1983.0',
 '1968.0',
 '1974.0',
 '2008.0',
 '1959.0',
 '2007.0',
 '1984.0',
 '1976.0',
 '1997.0',
 '1989.0',
 '1958.0',
 '1988.0',
 '2012.0',
 '1990.0',
 '1946.0',
 '1917.0',
 '2002.0',
 '2005.0',
 '2001.0',
 '1963.0',
 '1954.0',
 '1951.0',
 '1981.0',
 '1955.0',
 '1999.0',
 '2009.0',
 '1973.0',
 '1994.0',
 '1998.0',
 '1992.0',
 '1950.0',
 '1956.0',
 '2015.0',
 '0.0',
 '1932.0',
 '1.0',
 '1937.0',
 '1938.0',
 '1939.0',
 '1991.0',
 '1934.0',
 '1935.0',
 '20052009.0',
 '1947.0',
 '1953.0',
 '1933.0',
 '2016.0',
 '1930.0',
 '1912.0',
 '1929.0',
 '3.0',
 '1928.0',
 '1915.0',
 '1936.0',
 '1925.0',
 '1940.0',
 '1943.0',
 '1927.0',
 '18

In [ ]:
def check(x):
  try:
    return(float(x))
  except:
    return 0

df['build_year'] = df['build_year'].apply(check)
df['build_year'].unique()

array([1.9070000e+03, 1.9800000e+03, 2.0140000e+03, 1.9700000e+03,
       1.9820000e+03, 2.0130000e+03, 2.0040000e+03, 2.0030000e+03,
       1.9570000e+03, 1.9860000e+03, 1.9600000e+03, 1.9950000e+03,
       1.9790000e+03, 1.9750000e+03, 1.9870000e+03, 1.9620000e+03,
       1.9690000e+03, 1.9930000e+03, 1.9960000e+03, 1.9720000e+03,
       2.0110000e+03, 1.9650000e+03, 2.0100000e+03, 1.9850000e+03,
       2.0060000e+03, 1.9610000e+03, 1.9710000e+03, 1.9780000e+03,
       1.9660000e+03, 1.9670000e+03, 2.0000000e+03, 1.9640000e+03,
       1.9770000e+03, 1.9830000e+03, 1.9680000e+03, 1.9740000e+03,
       2.0080000e+03, 1.9590000e+03, 2.0070000e+03, 1.9840000e+03,
       1.9760000e+03, 1.9970000e+03, 1.9890000e+03, 1.9580000e+03,
       1.9880000e+03, 2.0120000e+03, 1.9900000e+03, 1.9460000e+03,
       1.9170000e+03, 2.0020000e+03, 2.0050000e+03, 2.0010000e+03,
       1.9630000e+03, 1.9540000e+03, 1.9510000e+03, 1.9810000e+03,
       1.9550000e+03, 1.9990000e+03, 2.0090000e+03, 1.9730000e

In [ ]:
df['build_year'] = df['build_year'].fillna(df['build_year'].median())

In [ ]:
for i in range(0, 18866, 1):
  if df['build_year'].loc[i] <= 1800:
    df['build_year'].loc[i] = df['build_year'].median()
df['build_year'].unique()

array([1907., 1980., 2014., 1970., 1982., 2013., 2004., 2003., 1957.,
       1986., 1960., 1995., 1979., 1975., 1987., 1962., 1969., 1993.,
       1996., 1972., 2011., 1965., 2010., 1985., 2006., 1961., 1971.,
       1978., 1966., 1967., 2000., 1964., 1977., 1983., 1968., 1974.,
       2008., 1959., 2007., 1984., 1976., 1997., 1989., 1958., 1988.,
       2012., 1990., 1946., 1917., 2002., 2005., 2001., 1963., 1954.,
       1951., 1981., 1955., 1999., 2009., 1973., 1994., 1998., 1992.,
       1950., 1956., 2015., 1932., 1937., 1938., 1939., 1991., 1934.,
       1935., 1947., 1953., 1933., 2016., 1930., 1912., 1929., 1928.,
       1915., 1936., 1925., 1940., 1943., 1927., 1896., 1911., 1924.,
       1952., 2017., 1926., 1931., 1860., 1949., 1914., 1910., 1895.,
       1948., 1876., 1900., 1890., 1920., 1904., 1906., 1941., 1905.,
       1886., 2018.])

In [ ]:
df[df['build_year']>=2021]

id   timestamp  ...  rent_price_1room_eco  average_life_exp
482   10092  2013-08-27  ...                 33.15             76.37
4614  15223  2014-02-19  ...                 33.65             76.70

[2 rows x 20 columns]

In [ ]:
df['build_year'].loc[482] = 2005
df['build_year'].loc[4614] = 1965

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


# num_room

In [ ]:
df['num_room'].unique()

array([1., 2., 3., 4., 5., 6., 8., 7., 9.])

In [ ]:
df[df['num_room']>=5]

In [ ]:
df['num_room'] = df['num_room'].fillna(0)

In [ ]:
for i in range(0, 18866, 1):
  if df['num_room'].loc[i] == 0:
    df['num_room'].loc[i] = df['num_room'].median()
  elif df['num_room'].loc[i]>=5 and df['full_sq'].loc[i]<=80:
    df['num_room'].loc[i] = 4

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


# rent_price_1room_eco

In [ ]:
df['rent_price_1room_eco'].unique()

array([32.31, 32.23, 32.55, 33.15, 34.53, 34.52, 34.55, 33.97, 33.49,
       33.65, 33.26, 32.96, 32.87, 33.06, 33.67, 34.83, 35.37, 34.3 ,
       34.33, 33.88, 34.35, 31.35, 31.41, 31.09, 30.29, 31.1 , 32.11,
       31.53, 33.1 , 31.11, 32.69, 32.93, 32.9 , 30.67, 33.07, 31.64,
       30.48, 32.62, 30.74, 32.61, 31.12, 32.84, 31.25, 30.89, 30.63])

In [ ]:
# valss = [i for i in df['rent_price_1room_eco'].unique()]
# valss

In [ ]:
for i in range(0, 18866, 1):
  if df['rent_price_1room_eco'].loc[i] == 2.31:
    df['rent_price_1room_eco'].loc[i] = 32.31

# rent_price_2room_eco

In [ ]:
df['rent_price_2room_eco'].unique()

array([40.77, 41.09, 41.42, 42.33, 42.1 , 41.96, 42.02, 42.  , 41.44,
       41.79, 41.91, 40.95, 41.17, 41.18, 41.97, 43.11, 43.26, 43.85,
       43.23, 42.64, 43.29, 41.86, 40.49, 39.92, 39.16, 38.91, 38.25,
       40.22, 38.62, 41.01, 38.1 , 41.19, 40.83, 40.66, 37.96, 38.45,
       39.  , 40.1 , 37.93, 41.3 , 38.39, 40.25, 39.47, 37.85, 37.73])

In [ ]:
for i in range(0, 18866, 1):
  if df['rent_price_2room_eco'].loc[i] == 0.1:
    df['rent_price_2room_eco'].loc[i] = 40.1

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


# Отчет

### <center> Отчет по анализу:
    
- **id:** без изменений;
- **timestamp:** без изменений;
- **full_sq:** 5326 решил поменять на 326; все показатели, которые меньше 30, приравнял к 30; 
- **life_sq:** 
  * заполнил NaN нулями;
  * значения, которые меньше 15, приравнял к 15;
  * значения, которые больше общей площади, приравнял к разнице между общей площадью и площадью кухни (исправленной); 
- **floor:**
  * заполнил NaN нулями;
  * значения, которые равнялись нулю, приравнял к медиане данного столбца при условии, что она меньше показателя максимального этажа (исправленного) в данной строчке;
  * значения, которые были больше значения макисмального этажа (исправленного) в данной строчке, приравнял к максимальному этажу (исправленному) в данной строчке;
  * значения, которые равнялись нулю, приравнял к максимальному этажу (исправленному) в данной строчке при условии, что медиана данного столбца больше или равна показатель максимального этажа (исправленного) в данной строчке;  
- **max_floor:** 
  * заполнил NaN нулями;
  * значения, которые равнялись нулю, приравнял к медиане данного столбца;
  * для строчек, в которых floor был равен NaN на момент работы со столбцом max_floor, max_floor также приравнял к медиане данного столбца;
- **material:**  заполнил NaN медианой для данного столбца;
- **build_year:**
  * произведена замена "o" на нуль;
  * заполнил NaN медианой для данного столбца;
  * значения, которые были меньше или равны 1800, приравнял к медиане для данного столбца;
  * поменял 20052009 на 2005;
  * поменял 4965 на 1965;
- **num_room:** 
  * заполнил NaN нулями;
  * нулевые значения приравнял к медиане для данного столбца;
  * значения, которые больше или равны 5, приравнял к 4 при условии, что full_sq для данной строчки меньше или равен 80
- **kitch_sq:** 
  * заполнил NaN 15 из собственных соображений о площади кухни;
  * значения, которые меньше 15, приравнял к 15 (собственные соображения);
  * значения, которые больше 25, приравнял к 25 (собственные соображения);
- **full_all:** без изменений;
- **state:** 
  * заполнил NaN нулями;
  * нулевые значения приравнял к медиане для данного столбца;
  * 33 поменял на 3;
- **sub_area:** без изменений;
- **price_doc:** без изменений;
- **timestamp** : без изменений;
- **salary** : без изменений;
- **fixed_basket**: без изменений;
- **rent_price_3room_eco**: без изменений;
- **rent_price_2room_eco**: значения, равные 0.1, поменял на 40.1, проведя анализ уникальных значений для данного столбца;
- **rent_price_1room_eco**: значения, равные 2.31, поменял на 32.31, проведя анализ уникальных значений для данного столбца;
- **average_life_exp**: без изменений;